src/testRunner/unittests/transform.ts

```diff
@@ -101,17 +101,17 @@ namespace ts {
 
         testBaseline("types", () => {
             return transformSourceFile(`let a: () => void`, [
-                context => file => visitNode(file, function visitor(node: Node): VisitResult<Node> {
+                context => file => visitNode(file, function visitor<T extends Node>(node: T): VisitResult<T> {
                     return visitEachChild(node, visitor, context);
                 })
             ]);
         });
 
         testBaseline("transformDefiniteAssignmentAssertions", () => {
             return transformSourceFile(`let a!: () => void`, [
-                context => file => visitNode(file, function visitor(node: Node): VisitResult<Node> {
+                context => file => visitNode(file, function visitor<T extends Node>(node: T): VisitResult<T> {

```

Comment:
This idea that visitors need to be generic and return a `T`-ish thing seems to be consistent bogus.

Reply:
No, doesn't seem to be, actually. The problem with these calls it that they reuse the same visitor for the top level function (which should return SourceFile to maintain visitNode's invariants), but also for children, which can be different types.

Reply:
I guess the problem is that our visitor implies that we should only ever return things related to the input node. But below, this visitor in fact does not. It takes a `KeywordTypeNode<SyntaxKind.VoidKeyword>` and returns a `KeywordTypeNode<SyntaxKind.UndefinedKeyword>`, which _should_ be illegal in the API, but clearly, we intend for that to be okay.

I'm not sure what to do here. This comes down to the intent of the API, and how much we want to guide people into doing the right thing.

Reply:
At best I can imagine that an `Expression` or `Statement` can expand out to several because they can flatten out to sequential statements or comma expressions, or something like that - but otherwise I don't think it's a coherent expectation.

Reply:
Maybe it needs to be `<T extends Node = Node, U extends Node = Node>(node: T): VisitResult<U>` such that the APIs that are guaranteed to pass the return result out maintain that info?

It also gets more fun when you look at `visitNode`, which takes the visit result and either returns it, or if it's the array, calls `lift` and returns whatever that is. So, this can't be parameterized on `T` at all, because the impliciation is that you can return whatever array of nodes you want so long as you have a lift function that does the right thing.

Reply:
That also implies that `VisitResult<T> = T | T[] | undefined` is bogus, because the array is sometimes not related at all to the single-value'd result.

Reply:
Exactly, `lift` can handle some array of its own types.

Reply:
```
export type VisitResult<T extends Node, U extends Node = T> = T | U[] | undefined;
```

And propagate its uses 😢

Reply:
And we haven't even touched on `Transformer` :((((

Reply:
Another terrible example I found as I was debugging: 
![image](https://user-images.githubusercontent.com/5341706/180579079-2de060e3-568b-451a-a630-0a0ba0c382f1.png)

Reply:
Going to close this particular thread; I'm pretty sure I addressed these visitor issues.